subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -332,30 +410,121 @@ func (h *HivedAlgorithm) initFreeCellList() {
+					"VC %v would expect being allocated bad cells.",
+					h.cellTypes[chain][level], chain, level, vcName, h.totalRemainingQuotas[chain][level],
+					h.numLeftCells[chain][level]-h.numBadCells[chain][level], h.numBadCells[chain][level], vcName)
+				for _, vc := range h.vcSchedulers[vcName].getNonReservedFreeCellList()[chain][level] {
+					virtualCell := vc.(*VirtualCell)
+					if virtualCell.GetPhysicalCell() == nil && virtualCell.GetStatus().State != api.BadState {
+						virtualCell.GetStatus().State = api.BadState
+						break
+					}
+				}
+			}
+		}
+	}
+
+	if parent := c.GetParent(); parent != nil {
+		for _, buddy := range parent.GetChildren() {
+			if buddy.(*PhysicalCell).GetStatus().State != api.BadState {
+				return
+			}
+		}
+		h.setBad(parent.(*PhysicalCell))
+	}
+}
+
+// setHealthy marks a physical cell as healthy and recursively for its parent, guaranteeing that
+// a cell is healthy if any of its children is healthy. The virtual cell that this physical cell
+// is bound to will also be marked as healthy.
+func (h *HivedAlgorithm) setHealthy(c *PhysicalCell) {
+	newState := api.FreeState
+	if c.GetPriority() != freePriority {
+		newState = api.UsedState
+	}
+	c.GetStatus().State = newState
+	if c.virtualCell != nil {
+		c.virtualCell.GetStatus().State = newState
+		c.virtualCell.GetStatus().PhysicalCell.State = newState
+		c.GetStatus().VirtualCell.State = newState
+	}
+	chain, level := c.GetChain(), c.GetLevel()
+	h.numBadCells[chain][level]--
+	if h.totalRemainingQuotas[chain][level] >= h.numLeftCells[chain][level]-h.numBadCells[chain][level] {
+		if h.totalRemainingQuotas[chain][level] == h.numLeftCells[chain][level]-h.numBadCells[chain][level] {
+			klog.Infof("Cell type %v (chain %v level %v) now has sufficient healthy cells for "+
+				"satisfying the total remaining quota of the VCs (quota %v, healthy %v, bad %v).",
+				h.cellTypes[chain][level], chain, level, h.totalRemainingQuotas[chain][level],
+				h.numLeftCells[chain][level]-h.numBadCells[chain][level], h.numBadCells[chain][level])
+		}
+		for vcName, vcQuota := range h.vcRemainingQuotas {
+			if vcQuota[chain][level] >= h.numLeftCells[chain][level]-h.numBadCells[chain][level] {
+				if vcQuota[chain][level] == h.numLeftCells[chain][level]-h.numBadCells[chain][level] {
+					klog.Infof("Cell type %v (chain %v level %v) now has sufficient healthy cells for "+
+						"satisfying the remaining quota of the VC %v (quota %v, healthy %v, bad %v).",
+						h.cellTypes[chain][level], chain, level, vcName, h.totalRemainingQuotas[chain][level],
+						h.numLeftCells[chain][level]-h.numBadCells[chain][level], h.numBadCells[chain][level])
+				}
+				for _, vc := range h.vcSchedulers[vcName].getNonReservedFreeCellList()[chain][level] {
+					virtualCell := vc.(*VirtualCell)
+					if virtualCell.GetPhysicalCell() == nil && virtualCell.GetStatus().State == api.BadState {
+						virtualCell.GetStatus().State = api.FreeState
+						break
```

Comment:
If previous state is "used" and "bad", you should convert it to "used" instead of "free".

So we may need 2 Dimensions cellstate:
free/used
bad/heathy